Source/JavaScriptCore/interpreter/Interpreter.cpp

```diff
@@ -461,19 +461,48 @@ void Interpreter::getAsyncStackTrace(JSCell* owner, Vector<StackFrame>& results,
 
     VM& vm = this->vm();
 
-    auto getParentGenerator = [&](JSGenerator* gen) -> JSGenerator* {
-        JSValue contextValue = gen->internalField(static_cast<unsigned>(JSGenerator::Field::Context)).get();
-        JSPromise* awaitedPromise = jsDynamicCast<JSPromise*>(contextValue);
-        if (awaitedPromise && awaitedPromise->status(vm) == JSPromise::Status::Pending) {
-            JSValue reactionsValue = awaitedPromise->internalField(JSPromise::Field::ReactionsOrResult).get();
+    auto getContextValueFromPromise = [&](JSPromise* promise) -> JSValue {
+        if (promise && promise->status(vm) == JSPromise::Status::Pending) {
+            JSValue reactionsValue = promise->internalField(JSPromise::Field::ReactionsOrResult).get();
             if (JSObject* reactions = jsDynamicCast<JSObject*>(reactionsValue)) {
-                Structure* structure = reactions->structure();
-                unsigned attributes;
-                PropertyOffset offset = structure->getConcurrently(vm.propertyNames->builtinNames().contextPrivateName().impl(), attributes);
-                if (offset != invalidOffset && !(attributes & (PropertyAttribute::Accessor | PropertyAttribute::CustomAccessorOrValue))) {
-                    JSValue contextFieldValue = reactions->getDirect(offset);
-                    if (auto* resultGenerator = jsDynamicCast<JSGenerator*>(contextFieldValue))
-                        return resultGenerator;
+                Structure* reactionsStructure = reactions->structure();
+                unsigned contextFieldAttributes;
+                PropertyOffset contextOffset = reactionsStructure->getConcurrently(vm.propertyNames->builtinNames().contextPrivateName().impl(), contextFieldAttributes);
```

Comment:
might be worth making an InternalFieldObjectImpl<3> for this. I've seen `get` show up in profiling many times. And this is something that will get called often since the Error stack is materialized in finalizeUnconditionally.

Reply:
I didn't add any new mechanism, but I simply modified it to use InternalFieldObjectImpl instead. I think this makes it much less fragile now.